keps/sig-node/runtime-class-scheduling.md

```diff
@@ -138,13 +138,12 @@ The RuntimeClass definition is augmented with an optional `Topology` struct:
 
 ```go
 type Topology struct {
-    // nodeSelectorTerm selects the set of nodes that support this RuntimeClass.
-    // Pods using this RuntimeClass can only be scheduled to a node matched by this selector.
-    // The nodeSelectorTerm is merged with a pod's other node affinity match
-    // expressions by appending the additional requirements to each preexisting
-    // NodeSelectorTerm.
+    // nodeSelector selects the set of nodes that support this RuntimeClass.
+    // Pods using this RuntimeClass can only be scheduled to a node matched by
+    // this selector. The nodeSelector is intersected (AND) with a pod's other
+    // node affinity or node selector requirements.
     // +optional
-    NodeSelectorTerm []corev1.NodeSelectorRequirement
+    NodeSelector corev1.NodeSelector
 
     // tolerations adds tolerations to pods running with this RuntimeClass.
```

Comment:
I am a bit unsure about having tolerations as a part of "topology" in RuntimeClass. This is useful if one doesn't want to provide RuntimeClass on all their pods. If they provide RuntimeClass for all their pods they can just use NodeSelector. If they insist on tainting some nodes, they can use regular Toleration (which is not a part of RuntimeClass). So, I guess my question is "what is the major benefit of having tolerations" here?

Reply:
I think this is covered under the Tolerations discussion on line 170:

> **Tolerations**
> 
> While NodeSelectors and labels are used for steering pods _towards_ nodes,
> [taints and tolerations][taint-and-toleration] are used for steering pods _away_
> from nodes. If every pod had a RuntimeClass and every RuntimeClass had a strict
> NodeSelector, then RuntimeClasses could use non-overlapping selectors in place
> of taints & tolerations. However the same could be said of regular pod
> selectors, yet taints & tolerations are still a useful addition. Examples of
> [use cases](#user-stories) for including tolerations in RuntimeClass topology
> inculde:
> 
> - Tainting Windows nodes with `windows.microsoft.com` to keep default linux pods
>   away from the nodes. Windows RuntimeClasses would then need a corresponding
>   toleration.
> - Tainting "sandbox" nodes with `sandboxed.kubernetes.io` to keep services
>   providing privileged node features away from sandboxed workloads. Sandboxed
>   RuntimeClasses would need a toleration to enable them to be run on those
>   nodes.

I agree that it's not strictly necessary, but I think it will be useful in practice. I guess both of the use cases listed could be accomplished with a `default` RuntimeClass that looked like this:

```
kind: RuntimeClass
metadata:
  name: default
handler: runc
nodeSelector:
  nodeSelectorTerms:
    matchExpressions:
      - key: "windows.microsoft.com"
        operator: DoesNotExist
      - key: "sandboxed.kubernetes.io"
        operator: DoesNotExist
```

Reply:
Does this assume that the toleration key is also set as label ? I think that this isn't always the case.

Reply:
This implies using labels rather than taints.

Reply:
it is more staightforward for users to implement dedicated nodes using taints and tolerations rather than nodeSelectors. 
The benefit of having tolerations on the runtimeClass is freeing users that want to consume a specific runtimeClass available on a dedicated node to think about which toleration they should set or not.

Reply:
Your points are valid. Thanks for clarifying.